File: JavaDemo/src/main/java/com/lanayru/javademo/VolatileTest.java
package com.lanayru.javademo;

public class VolatileTest {

    long a = 0;
    boolean flag = false;

    private void write() {
        System.out.println("write in thread " + Thread.currentThread().getName());
        flag = true;
        a = 10000;
        System.out.println("end write in thread " + Thread.currentThread().getName());

        try {
            Thread.sleep(100);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        a = 20000;
    }

    private void read() {
        System.out.println("read in thread " + Thread.currentThread().getName());
        if (flag) {
            long i = a;
            System.out.println("a is " + i);
        }
    }

    public void test() {
        new Thread(new Runnable() {
            @Override
            public void run() {
                write();
            }
        }).start();

        new Thread(new Runnable() {
            @Override
            public void run() {
                read();
            }
        }).start();

        new Thread(new Runnable() {
            @Override
            public void run() {
                read();
            }
        }).start();

        try {
            Thread.sleep(200);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
}

File: app/src/test/java/com/lanayru/app/VolatileTest.kt
package com.lanayru.app

import org.junit.Test
import kotlin.concurrent.thread

class VolatileTest {

//        @Volatile var a = 0
    var a = 0
    var flag = false


    private fun write() {
        println("write in thread ${Thread.currentThread().name}")
        a += 1

        flag = true

        Thread.sleep(200)

        println("write end")
    }

    private fun read() {
//        println("i is $a")
        println("read in thread ${Thread.currentThread().name}")
        if (flag) {
//            val i = a
            println("i is $a")
        }
    }

    private fun execute(f: () -> Unit) {
        thread(start=true) {
            f()
        }
    }

    @Test
    fun testWR() {
        execute {
            write()
        }

        execute {
            write()
        }

        execute {
            read()
        }

        execute { read() }

        Thread.sleep(1000)
    }
}